packages/synapse-react-client/src/lib/utils/hooks/useLogin.ts

```diff
@@ -0,0 +1,176 @@
+   *
+   * If the user has 2FA enabled, the client app will receive a 401 with the userId and twoFaToken in the response.
+   *
+   * The app can pass the error response as a prop/argument, and we'll use it here to jump straight to the VERIFICATION_CODE step.
+   */
+  useEffect(() => {
+    if (twoFaErrorResponse) {
+      setTwoFaToken(twoFaErrorResponse.twoFaToken)
+      setUserId(twoFaErrorResponse.userId)
+      if (!['VERIFICATION_CODE', 'RECOVERY_CODE', 'LOGGED_IN'].includes(step)) {
+        setStep('VERIFICATION_CODE')
+      }
+    }
+  }, [twoFaErrorResponse])
+
+  async function finishLogin(loginResponse: LoginResponse) {
+    await SynapseClient.setAccessTokenCookie(loginResponse.accessToken)
+    localStorage.setItem(
+      AUTHENTICATION_RECEIPT_LOCALSTORAGE_KEY,
+      loginResponse.authenticationReceipt,
+    )
+    setStep('LOGGED_IN')
+
+    if (sessionCallback) {
+      sessionCallback()
+    }
+  }
+
+  const submitUsernameAndPassword: UseLoginReturn['submitUsernameAndPassword'] =
+    async (username, password) => {
+      setErrorMessage(undefined)
+      try {
+        const authenticationReceipt = localStorage.getItem(
+          AUTHENTICATION_RECEIPT_LOCALSTORAGE_KEY,
+        )
+        const data = await SynapseClient.login(
+          username,
+          password,
+          authenticationReceipt,
+        )
+        if ('errorCode' in data) {
+          setStep('VERIFICATION_CODE')
+          setTwoFaToken(data.twoFaToken)
+          setUserId(data.userId)
+        } else {
+          await finishLogin(data)
+        }
+      } catch (err) {
+        setErrorMessage(err.reason)
+      }
+    }
+
+  const submitOneTimePassword: UseLoginReturn['submitOneTimePassword'] = async (
+    code,
+    otpType = step === 'RECOVERY_CODE' ? 'RECOVERY_CODE' : 'TOTP',
+  ) => {
+    setErrorMessage(undefined)
+    if (twoFaToken == null || userId == null) {
+      // This could happen if the 2FA component exists on its own route, and the user directly navigates to it without first logging in with credentials/OAuth
+      setErrorMessage(
+        'You did not first log in with your password or a third-party identity provider.',
+      )
+      return
+    }
+    const request: TwoFactorAuthLoginRequest = {
+      userId: userId,
+      twoFaToken: twoFaToken,
+      otpCode: code,
+      otpType: otpType,
+    }
+    try {
+      const loginResponse = await SynapseClient.loginWith2fa(request)
+      await finishLogin(loginResponse)
+    } catch (e) {
+      setErrorMessage(e.reason)
+      if (
+        // The twoFaToken wasn't transmitted correctly
+        e.reason.includes('The provided twoFaToken is invalid') ||
+        // The user waited too long to enter the code.
+        e.reason.includes('Token has expired')
+      ) {
+        // Instruct the user refresh to start over.
+        setErrorMessage('Something went wrong. Refresh the page and try again.')
```

Comment:
If we are overwriting the error message in the UI, could you send the actual reason to the console as well?
It's possible that the exact strings used ("The provided twoFaToken is invalid") could change over time, but the consequence is minimal (the user would see a confusing error), so I'm fine with this